frontend/components/AddFundsThroughBridge/useGenerateAddFunds.tsx

```diff
@@ -0,0 +1,126 @@
+ */
+export const useGenerateAddFunds = ({
+  requirements,
+  defaultTokenAmounts,
+  destinationAddress,
+  onlyNativeToken = false,
+}: {
+  requirements: AddressBalanceRecord;
+  defaultTokenAmounts?: DefaultTokenAmount[];
+  destinationAddress?: Address;
+  onlyNativeToken?: boolean;
+}) => {
+  const { selectedAgentConfig } = useServices();
+  const toMiddlewareChain = selectedAgentConfig.middlewareHomeChainId;
+  const getBridgeRequirementsParams =
+    useAddFundsGetBridgeRequirementsParams(destinationAddress);
+
+  // In general, master_safe and master_eoa addresses are used.
+  const allAddresses = typedKeys(requirements);
+
+  // All token addresses that can be used to add funds to the master safe.
+  const tokenAddresses = allAddresses.reduce<Address[]>((acc, address) => {
+    const tokens = requirements[address];
+    if (!tokens) return acc;
+
+    // If onlyNativeToken is true, filter to include only the native token.
+    const typedKeysTokens = typedKeys(tokens).filter((token) =>
+      onlyNativeToken ? token === AddressZero : true,
+    );
+    acc.push(...typedKeysTokens);
+    return acc;
+  }, []);
+
+  const amountsToReceive: GeneratedInput[] = uniq(tokenAddresses).map(
+    (tokenAddress: Address) => {
+      const symbol = getTokenDetailsFromAddress(
+        toMiddlewareChain,
+        tokenAddress,
+      ).symbol;
+
+      // if no default token amounts are provided, set amount to 0.
+      // user can update the amount later.
+      const defaultAmount =
+        defaultTokenAmounts?.find((token) => token.symbol === symbol)?.amount ??
+        0;
+      return { tokenAddress, symbol, amount: defaultAmount };
+    },
+  );
+
+  const [inputs, setInputs] = useState(
+    amountsToReceive.reduce(
+      (acc, { symbol, amount }) => ({ ...acc, [symbol]: amount }),
+      {} as Record<TokenSymbol, number | null>,
+    ),
+  );
+
+  const handleInputChange = useCallback(
+    (symbol: TokenSymbol, value: number | null) => {
+      setInputs((prev) => ({ ...prev, [symbol]: value }));
+    },
+    [],
+  );
+
+  // Generate bridge requirements params for the inputs provided by the user.
+  const bridgeRequirementsParams = useMemo(() => {
+    return amountsToReceive.reduce<BridgeRequest[]>((acc, tokenDetails) => {
+      const amount = inputs[tokenDetails.symbol];
+      if (!amount) return acc;
+
+      const param = getBridgeRequirementsParams(
+        tokenDetails.tokenAddress,
+        amount,
+      );
+      acc.push(param);
+      return acc;
+    }, []);
+  }, [amountsToReceive, inputs, getBridgeRequirementsParams]);
+
+  // If all inputs are empty or zero, disable the button
+  const isInputEmpty = useMemo(
+    () => Object.values(inputs).every((v) => !v || v <= 0),
+    [inputs],
+  );
+
+  const inputsToDisplay = useMemo(
+    () =>
+      Object.entries(inputs).map(([symbol, amount]) => ({
+        symbol: symbol as TokenSymbol,
+        amount: amount ?? 0,
+      })),
+    [inputs],
```

Comment:
can be in one useMemo, since has the same dependancy.

Reply:
```
 const { isInputEmpty, inputsToDisplay } = useMemo(() => {
    const isInputEmpty = Object.values(inputs).every((v) => !v || v <= 0);
    const inputsToDisplay = Object.entries(inputs).map(([symbol, amount]) => ({
      symbol: symbol as TokenSymbol,
      amount: amount ?? 0,
    }));

    return { isInputEmpty, inputsToDisplay };
  }, [inputs]);

```

Reply:
ok, I'm generally against memoization as such without deep investigation if it's really needed, cuz it's not free at all, but up to you